src/shared/sd-dns-resolver.c

```diff
@@ -0,0 +1,362 @@
+        uint16_t port = 0;
+        _cleanup_free_ char *dohpath = NULL;
+        bool alpn = false;
+
+        uint16_t lastkey = 0;
+        while (offset < len) {
+                if (offset + 4 > len)
+                        return -EBADMSG;
+
+                uint16_t key = unaligned_read_be16(&option[offset]);
+                offset += 2;
+
+                /* RFC9460 § 2.2 SvcParam MUST appear in strictly increasing numeric order */
+                if (lastkey >= key)
+                        return -EBADMSG;
+                lastkey = key;
+
+                uint16_t plen = unaligned_read_be16(&option[offset]);
+                offset += 2;
+                if (offset + plen > len)
+                        return -EBADMSG;
+
+                switch (key) {
+                /* Mandatory keys must be understood by the client, otherwise the record should be discarded.
+                 * Automatic mandatory keys must not appear in the mandatory parameter, so these are all
+                 * supplementary. We don't understand any supplementary keys, so if the mandatory parameter
+                 * is present, we cannot use this record.*/
+                case DNS_SVC_PARAM_KEY_MANDATORY:
+                        if (plen > 0)
+                                return -EBADMSG;
+                        break;
+
+                case DNS_SVC_PARAM_KEY_ALPN:
+                        if (plen == 0)
+                                return 0;
+                        alpn = true; /* alpn is required. Record that the requirement is met. */
+
+                        size_t poff = offset;
+                        size_t pend = offset + plen;
+                        while (poff < pend) {
+                                uint8_t alen = option[poff++];
+                                if (poff + alen > len)
+                                        return -EBADMSG;
+                                if (memcmp_nn((const char*) &option[poff], alen, "dot", STRLEN("dot")) == 0)
+                                        SET_FLAG(transports, SD_DNS_ALPN_DOT, true);
+                                if (memcmp_nn((const char*) &option[poff], alen, "h2", STRLEN("h2")) == 0)
+                                        SET_FLAG(transports, SD_DNS_ALPN_HTTP_2_TLS, true);
+                                if (memcmp_nn((const char*) &option[poff], alen, "h3", STRLEN("h3")) == 0)
+                                        SET_FLAG(transports, SD_DNS_ALPN_HTTP_3, true);
+                                if (memcmp_nn((const char*) &option[poff], alen, "doq", STRLEN("doq")) == 0)
+                                        SET_FLAG(transports, SD_DNS_ALPN_DOQ, true);
+                                poff += alen;
+                        }
+                        if (poff != pend)
+                                return -EBADMSG;
+                        break;
+
+                case DNS_SVC_PARAM_KEY_PORT:
+                        if (plen != 2)
```

Comment:
doesn't matter but I'd usually write the 2 as `sizeof(uint16_t)` to make the intention clear here...